zerver/tornado/event_queue.py

```diff
@@ -1182,6 +1187,23 @@ def process_message_update_event(
                 client.add_event(user_event)
 
 
+def process_custom_profile_fields_event(event: Mapping[str, Any], users: Iterable[int]) -> None:
+    pronouns_type_fields = [
+        field for field in event["fields"] if field["type"] == CustomProfileField.PRONOUNS
+    ]
+
+    for user_profile_id in users:
+        for client in get_client_descriptors_for_user(user_profile_id):
+            if client.accepts_event(event):
+                if not client.pronouns_field_type_supported:
+                    field_type = CustomProfileField.SHORT_TEXT
+                else:
+                    field_type = CustomProfileField.PRONOUNS
+                for field in pronouns_type_fields:
+                    field["type"] = field_type
+                client.add_event(event)
```

Comment:
It looks like this mutates the event dict, then passes it to `client.add_event`, then proceeds to the next client and mutates the same event dict that it already passed to `client.add_event`.

Is that the usual approach in this file? Out of context, it makes me a bit nervous -- it's the sort of pattern that often means a bug.

The way I'd be inclined to write this would be:
* If the client does support the field type, then just pass the event through untouched. (This should be the common case, and also the case the reader should focus on first to understand the function.)
* If it doesn't, then construct a new event dict -- leaving the original one unmutated, to potentially use for the next client -- with the adjustment, and pass that.
  * As an optional bonus, perhaps memoize that new adjusted event dict in case there are many clients that need it.

Reply:
Good catch -- this is indeed a bug.

Reply:
I think we can just construct a `pronouns_field_unsupported_event` object at the top of the function, and just have the logic decide which event to send to each client.

The `slim_presence` event code path has an example of this pattern.